File: admin-mit-backend/app/services/system.py
"""
系统设置服务
"""
import os
import platform
import time
from datetime import datetime, timedelta
from flask import g
from app.extensions import db
from app.models.system import SystemSetting
from app.models.user import User

# 尝试导入psutil，如果不存在则使用备用方案
try:
    import psutil
    HAS_PSUTIL = True
except ImportError:
    HAS_PSUTIL = False


class SystemService:
    """系统设置服务类"""
    
    # 默认系统配置
    DEFAULT_SETTINGS = {
        'system': {
            'system_name': 'MiTong运维管理系统',
            'system_version': '2.1.0',
            'system_description': '企业级运维管理平台，提供主机管理、监控告警、网络探测等功能',
            'company_name': '米通科技有限公司',
            'company_logo': None,
            'timezone': 'Asia/Shanghai',
            'language': 'zh-CN',
            'maintenance_mode': False,
            'maintenance_message': '系统正在维护中，预计30分钟后恢复正常'
        },
        'security': {
            'max_login_attempts': 5,
            'session_timeout': 7200,  # 2小时
            'password_policy': {
                'min_length': 8,
                'require_uppercase': True,
                'require_lowercase': True,
                'require_numbers': True,
                'require_symbols': False,
                'expiry_days': 90
            },
            'audit_log_retention_days': 90  # 审计日志保留天数，默认90天
        },
        'backup': {
            'auto_backup': True,
            'backup_interval': 24,  # 小时
            'backup_retention': 30,  # 天
            'backup_location': '/var/backups/mitong'
        },
        'notification': {
            'email_enabled': True,
            'sms_enabled': False,
            'system_notifications': True,
            'email_config': {
                'smtp_host': '',
                'smtp_port': 587,
                'smtp_user': '',
                'smtp_password': '',
                'use_tls': True
            }
        }
    }
    
    @classmethod
    def initialize_default_settings(cls, tenant_id, user_id):
        """初始化默认系统设置"""
        try:
            for category, settings in cls.DEFAULT_SETTINGS.items():
                for key, value in settings.items():
                    setting_key = f"{category}.{key}"
                    
                    # 检查设置是否已存在
                    existing = SystemSetting.query.filter(
                        SystemSetting.tenant_id == tenant_id,
                        SystemSetting.key == setting_key
                    ).first()
                    
                    if not existing:
                        setting = SystemSetting(
                            tenant_id=tenant_id,
                            key=setting_key,
                            value=value,
                            category=category,
                            description=f"系统默认{category}设置",
                            is_system=True,
                            created_by=user_id
                        )
                        db.session.add(setting)
            
            db.session.commit()
            return True
        except Exception as e:
            db.session.rollback()
            raise e
    
    @classmethod
    def get_system_config(cls, tenant_id=None):
        """获取系统配置"""
        if tenant_id is None:
            tenant_id = getattr(g, 'tenant_id', None)
        
        if tenant_id is None:
            raise ValueError("租户ID不能为空")
        
        # 获取所有系统设置
        settings = SystemSetting.query.filter(
            SystemSetting.tenant_id == tenant_id,
            SystemSetting.is_enabled == True
        ).all()
        
        # 按分类组织配置
        config = {}
        for setting in settings:
            category, key = setting.key.split('.', 1) if '.' in setting.key else ('general', setting.key)
            
            if category not in config:
                config[category] = {}
            
            config[category][key] = setting.value
        
        # 合并默认配置（如果某些设置不存在）
        for category, default_settings in cls.DEFAULT_SETTINGS.items():
            if category not in config:
                config[category] = {}
            
            for key, default_value in default_settings.items():
                if key not in config[category]:
                    config[category][key] = default_value
        
        return config
    
    @classmethod
    def update_system_config(cls, config_data, tenant_id=None, user_id=None):
        """更新系统配置"""
        if tenant_id is None:
            tenant_id = getattr(g, 'tenant_id', None)
        if user_id is None:
            user_id = getattr(g, 'user_id', None)
        
        if tenant_id is None or user_id is None:
            raise ValueError("租户ID和用户ID不能为空")
        
        try:
            for category, settings in config_data.items():
                for key, value in settings.items():
                    setting_key = f"{category}.{key}"
                    
                    # 查找现有设置
                    setting = SystemSetting.query.filter(
                        SystemSetting.tenant_id == tenant_id,
                        SystemSetting.key == setting_key
                    ).first()
                    
                    if setting:
                        # 更新现有设置
                        setting.value = value
                        setting.updated_by = user_id
                        setting.updated_at = datetime.utcnow()
                    else:
                        # 创建新设置
                        setting = SystemSetting(
                            tenant_id=tenant_id,
                            key=setting_key,
                            value=value,
                            category=category,
                            description=f"用户自定义{category}设置",
                            created_by=user_id
                        )
                        db.session.add(setting)
            
            db.session.commit()
            return True
        except Exception as e:
            db.session.rollback()
            raise e
    
    @classmethod
    def get_system_info(cls):
        """获取系统信息"""
        try:
            # 获取系统基本信息
            system_info = {
                'hostname': platform.node(),
                'os_version': f"{platform.system()} {platform.release()}",
                'python_version': platform.python_version(),
                'cpu_cores': os.cpu_count() or 1,
                'memory_total': cls._get_memory_info(),
                'disk_total': cls._get_disk_info(),
                'uptime': cls._get_uptime(),
                'load_average': cls._get_load_average()
            }
            
            # 获取用户统计
            tenant_id = getattr(g, 'tenant_id', None)
            if tenant_id:
                total_users = User.query_by_tenant(tenant_id).count()
                # 简化活跃用户统计（这里可以根据实际需求调整）
                active_users = User.query_by_tenant(tenant_id).filter(User.status == 1).count()
                
                system_info.update({
                    'total_users': total_users,
                    'active_users': active_users
                })
            else:
                system_info.update({
                    'total_users': 0,
                    'active_users': 0
                })
            
            return system_info
        except Exception as e:
            # 返回默认信息
            return {
                'hostname': 'unknown',
                'os_version': 'unknown',
                'python_version': platform.python_version(),
                'cpu_cores': 1,
                'memory_total': '0 GB',
                'disk_total': '0 GB',
                'uptime': '0天 0小时 0分钟',
                'load_average': [0.0, 0.0, 0.0],
                'total_users': 0,
                'active_users': 0
            }
    
    @classmethod
    def _get_memory_info(cls):
        """获取内存信息"""
        if HAS_PSUTIL:
            try:
                return cls._format_bytes(psutil.virtual_memory().total)
            except:
                pass
        return "未知"
    
    @classmethod
    def _get_disk_info(cls):
        """获取磁盘信息"""
        if HAS_PSUTIL:
            try:
                # Windows使用C:，Linux使用/
                path = 'C:' if platform.system() == 'Windows' else '/'
                return cls._format_bytes(psutil.disk_usage(path).total)
            except:
                pass
        return "未知"
    
    @classmethod
    def _get_load_average(cls):
        """获取负载平均值"""
        if hasattr(os, 'getloadavg'):
            try:
                return list(os.getloadavg())
            except:
                pass
        return [0.0, 0.0, 0.0]
    
    @classmethod
    def _format_bytes(cls, bytes_value):
        """格式化字节数"""
        for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
            if bytes_value < 1024.0:
                return f"{bytes_value:.1f} {unit}"
            bytes_value /= 1024.0
        return f"{bytes_value:.1f} PB"
    
    @classmethod
    def _get_uptime(cls):
        """获取系统运行时间"""
        if HAS_PSUTIL:
            try:
                boot_time = psutil.boot_time()
                uptime_seconds = time.time() - boot_time
                uptime_delta = timedelta(seconds=uptime_seconds)
                
                days = uptime_delta.days
                hours, remainder = divmod(uptime_delta.seconds, 3600)
                minutes, _ = divmod(remainder, 60)
                
                return f"{days}天 {hours}小时 {minutes}分钟"
            except:
                pass
        return "未知"
    
    @classmethod
    def get_maintenance_status(cls, tenant_id=None):
        """获取维护模式状态"""
        maintenance_mode = SystemSetting.get_setting('system.maintenance_mode', tenant_id, False)
        maintenance_message = SystemSetting.get_setting('system.maintenance_message', tenant_id, '系统正在维护中')
        
        return {
            'maintenance_mode': maintenance_mode,
            'maintenance_message': maintenance_message
        }
    
    @classmethod
    def set_maintenance_mode(cls, enabled, message=None, tenant_id=None, user_id=None):
        """设置维护模式"""
        if tenant_id is None:
            tenant_id = getattr(g, 'tenant_id', None)
        if user_id is None:
            user_id = getattr(g, 'user_id', None)
        
        try:
            # 设置维护模式状态
            SystemSetting.set_setting(
                'system.maintenance_mode',
                enabled,
                'system',
                '维护模式开关',
                user_id
            )
            
            # 设置维护消息
            if message:
                SystemSetting.set_setting(
                    'system.maintenance_message',
                    message,
                    'system',
                    '维护模式提示消息',
                    user_id
                )
            
            return True
        except Exception as e:
            raise e
    
    @classmethod
    def export_settings(cls, tenant_id=None):
        """导出系统设置"""
        if tenant_id is None:
            tenant_id = getattr(g, 'tenant_id', None)
        
        if tenant_id is None:
            raise ValueError("租户ID不能为空")
        
        settings = SystemSetting.query.filter(
            SystemSetting.tenant_id == tenant_id,
            SystemSetting.is_enabled == True
        ).all()
        
        export_data = {
            'export_time': datetime.utcnow().isoformat(),
            'tenant_id': tenant_id,
            'settings': {}
        }
        
        for setting in settings:
            export_data['settings'][setting.key] = {
                'value': setting.value,
                'category': setting.category,
                'description': setting.description,
                'is_system': setting.is_system
            }
        
        return export_data
    
    @classmethod
    def import_settings(cls, import_data, tenant_id=None, user_id=None):
        """导入系统设置"""
        if tenant_id is None:
            tenant_id = getattr(g, 'tenant_id', None)
        if user_id is None:
            user_id = getattr(g, 'user_id', None)
        
        if tenant_id is None or user_id is None:
            raise ValueError("租户ID和用户ID不能为空")
        
        try:
            settings_data = import_data.get('settings', {})
            
            for key, setting_info in settings_data.items():
                # 跳过系统级设置的导入（安全考虑）
                if setting_info.get('is_system', False):
                    continue
                
                SystemSetting.set_setting(
                    key,
                    setting_info['value'],
                    setting_info.get('category', 'general'),
                    setting_info.get('description', '导入的设置'),
                    user_id
                )
            
            return True
        except Exception as e:
            raise e

File: admin-mit-ui/src/services/system.ts
/**
 * 系统设置 API 服务
 */
import { api } from './api'

// 系统配置接口
export interface SystemConfig {
  system: {
    system_name: string
    system_version: string
    system_description: string
    company_name: string
    company_logo?: string
    timezone: string
    language: string
    maintenance_mode: boolean
    maintenance_message: string
  }
  security: {
    max_login_attempts: number
    session_timeout: number
    password_policy: {
      min_length: number
      require_uppercase: boolean
      require_lowercase: boolean
      require_numbers: boolean
      require_symbols: boolean
      expiry_days: number
    }
  }
  backup: {
    auto_backup: boolean
    backup_interval: number
    backup_retention: number
    backup_location: string
  }
  notification: {
    email_enabled: boolean
    sms_enabled: boolean
    system_notifications: boolean
    email_config: {
      smtp_host: string
      smtp_port: number
      smtp_user: string
      smtp_password: string
      use_tls: boolean
    }
  }
}

// 系统信息接口
export interface SystemInfo {
  hostname: string
  os_version: string
  python_version: string
  cpu_cores: number
  memory_total: string
  disk_total: string
  uptime: string
  load_average: number[]
  total_users: number
  active_users: number
}

// 维护模式状态接口
export interface MaintenanceStatus {
  maintenance_mode: boolean
  maintenance_message: string
}

// 系统设置项接口
export interface SystemSetting {
  id: number
  key: string
  value: any
  category: string
  description?: string
  is_system: boolean
  is_enabled: boolean
  created_by: number
  updated_by?: number
  created_at: string
  updated_at: string
  creator_name?: string
  updater_name?: string
}

// 设置列表响应接口
export interface SettingsListResponse {
  settings: SystemSetting[]
  pagination: {
    page: number
    per_page: number
    total: number
    pages: number
    has_prev: boolean
    has_next: boolean
  }
}

// 导出数据接口
export interface ExportData {
  export_time: string
  tenant_id: number
  settings: Record<string, {
    value: any
    category: string
    description?: string
    is_system: boolean
  }>
}

// 安全统计接口
export interface SecurityStats {
  failed_logins_today: number
  locked_accounts: number
  active_sessions: number
  password_expiring_soon: number
  security_events_today: number
  last_security_scan: string | null
}

class SystemService {
  /**
   * 获取系统配置
   */
  async getSystemConfig(): Promise<SystemConfig> {
    const response = await api.get('/api/system/config')
    return response.data
  }

  /**
   * 更新系统配置
   */
  async updateSystemConfig(config: Partial<SystemConfig>): Promise<void> {
    await api.post('/api/system/config', config)
  }

  /**
   * 获取系统信息
   */
  async getSystemInfo(): Promise<SystemInfo> {
    const response = await api.get('/api/system/info')
    return response.data
  }

  /**
   * 获取维护模式状态
   */
  async getMaintenanceStatus(): Promise<MaintenanceStatus> {
    const response = await api.get('/api/system/maintenance')
    return response.data
  }

  /**
   * 设置维护模式
   */
  async setMaintenanceMode(enabled: boolean, message?: string): Promise<void> {
    await api.put('/api/system/maintenance', {
      enabled,
      message
    })
  }

  /**
   * 获取系统设置列表
   */
  async getSettings(params?: {
    category?: string
    page?: number
    per_page?: number
  }): Promise<SettingsListResponse> {
    const response = await api.get('/api/system/settings', { params })
    return response.data
  }

  /**
   * 获取单个设置
   */
  async getSetting(key: string): Promise<SystemSetting> {
    const response = await api.get(`/api/system/settings/${key}`)
    return response.data
  }

  /**
   * 更新单个设置
   */
  async updateSetting(key: string, data: {
    value: any
    description?: string
  }): Promise<SystemSetting> {
    const response = await api.put(`/api/system/settings/${key}`, data)
    return response.data
  }

  /**
   * 导出系统设置
   */
  async exportSettings(): Promise<ExportData> {
    const response = await api.get('/api/system/settings/export')
    return response.data
  }

  /**
   * 导入系统设置
   */
  async importSettings(data: ExportData): Promise<void> {
    await api.post('/api/system/settings/import', data)
  }

  /**
   * 初始化默认设置
   */
  async initializeDefaultSettings(): Promise<void> {
    await api.post('/api/system/initialize')
  }

  /**
   * 获取安全统计数据
   */
  async getSecurityStats(): Promise<SecurityStats> {
    const response = await api.get('/api/system/security/stats')
    return response.data
  }

  /**
   * 解锁所有被锁定的账户
   */
  async unlockAllAccounts(): Promise<{ unlocked_count: number; unlocked_users: string[] }> {
    const response = await api.post('/api/system/security/unlock-all')
    return response.data
  }

  /**
   * 下载导出文件
   */
  downloadExportFile(data: ExportData, filename?: string): void {
    const blob = new Blob([JSON.stringify(data, null, 2)], {
      type: 'application/json'
    })
    
    const url = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = url
    link.download = filename || `system-settings-${new Date().toISOString().split('T')[0]}.json`
    
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    
    window.URL.revokeObjectURL(url)
  }

  /**
   * 读取导入文件
   */
  readImportFile(file: File): Promise<ExportData> {
    return new Promise((resolve, reject) => {
      const reader = new FileReader()
      
      reader.onload = (e) => {
        try {
          const data = JSON.parse(e.target?.result as string)
          resolve(data)
        } catch (error) {
          reject(new Error('文件格式错误，请选择有效的JSON文件'))
        }
      }
      
      reader.onerror = () => {
        reject(new Error('文件读取失败'))
      }
      
      reader.readAsText(file)
    })
  }

  /**
   * 验证导入数据格式
   */
  validateImportData(data: any): boolean {
    if (!data || typeof data !== 'object') {
      return false
    }

    // 检查必需字段
    if (!data.settings || typeof data.settings !== 'object') {
      return false
    }

    // 检查设置格式
    for (const [key, setting] of Object.entries(data.settings)) {
      if (!setting || typeof setting !== 'object') {
        return false
      }

      const settingObj = setting as any
      if (!settingObj.hasOwnProperty('value') || 
          !settingObj.hasOwnProperty('category')) {
        return false
      }
    }

    return true
  }

  /**
   * 格式化字节数
   */
  formatBytes(bytes: number): string {
    if (bytes === 0) return '0 B'
    
    const k = 1024
    const sizes = ['B', 'KB', 'MB', 'GB', 'TB']
    const i = Math.floor(Math.log(bytes) / Math.log(k))
    
    return parseFloat((bytes / Math.pow(k, i)).toFixed(1)) + ' ' + sizes[i]
  }

  /**
   * 格式化运行时间
   */
  formatUptime(uptime: string): string {
    // 如果已经是格式化的字符串，直接返回
    if (uptime.includes('天') || uptime.includes('小时') || uptime.includes('分钟')) {
      return uptime
    }

    // 如果是秒数，进行格式化
    const seconds = parseInt(uptime)
    if (isNaN(seconds)) {
      return uptime
    }

    const days = Math.floor(seconds / 86400)
    const hours = Math.floor((seconds % 86400) / 3600)
    const minutes = Math.floor((seconds % 3600) / 60)

    return `${days}天 ${hours}小时 ${minutes}分钟`
  }

  /**
   * 获取系统状态颜色
   */
  getSystemStatusColor(value: number, thresholds: { warning: number; danger: number }): string {
    if (value >= thresholds.danger) {
      return 'text-red-600'
    } else if (value >= thresholds.warning) {
      return 'text-yellow-600'
    } else {
      return 'text-green-600'
    }
  }

  /**
   * 获取负载平均值状态
   */
  getLoadAverageStatus(loadAverage: number[], cpuCores: number): {
    status: 'good' | 'warning' | 'danger'
    color: string
    description: string
  } {
    const load1min = loadAverage[0] || 0
    const loadPercentage = (load1min / cpuCores) * 100

    if (loadPercentage < 70) {
      return {
        status: 'good',
        color: 'text-green-600',
        description: '系统负载正常'
      }
    } else if (loadPercentage < 90) {
      return {
        status: 'warning',
        color: 'text-yellow-600',
        description: '系统负载较高'
      }
    } else {
      return {
        status: 'danger',
        color: 'text-red-600',
        description: '系统负载过高'
      }
    }
  }
}

export const systemService = new SystemService()
export default systemService